File: app/src/main/java/cz/kutner/comicsdb/fragment/AbstractFragment.java
package cz.kutner.comicsdb.fragment;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.SearchView;

import java.util.ArrayList;
import java.util.List;

import cz.kutner.comicsdb.Utils;
import cz.kutner.comicsdb.event.AbstractResultEvent;
import cz.kutner.comicsdb.event.EventBus;
import cz.kutner.comicsdbclient.comicsdbclient.R;

/**
 * Created by Lukáš Kutner (lukas@kutner.cz) on 21.5.2015.
 */
public abstract class AbstractFragment<Item, Adapter extends RecyclerView.Adapter, Event extends AbstractResultEvent> extends Fragment {
    String LOG_TAG = getClass().getSimpleName();
    ViewGroup container;
    int lastPage;
    boolean firstLoad;
    boolean searchRunning;
    boolean loading;
    Item lastItem;
    List<Item> data = new ArrayList<>();
    List<Item> result = new ArrayList<>();
    int pastVisibleItems, visibleItemCount, totalItemCount;
    LinearLayoutManager llm;
    Adapter adapter;
    int fragment_view;
    int recycler_view;
    int preloadCount;
    boolean endless;

    public AbstractFragment() {
        EventBus.getInstance().register(this);
        lastPage = 1;
        loading = false;
        endless = true;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView;
        this.container = container;
        //TODO  - tohle kontrolovat pokaždé při načtení - v onStart?
        if (!Utils.isConnected(this.getActivity())) {
            rootView = inflater.inflate(R.layout.loading_error, container, false);
            Button tryAgainButton = (Button) rootView.findViewById(R.id.try_again);
            tryAgainButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (Utils.isConnected(getActivity())) {
                        loadData();
                    }
                }
            });
        } else {
            rootView = inflater.inflate(R.layout.loading, container, false);
        }
        return rootView;
    }

    abstract void loadData();

    @Override
    public void onDestroy() {
        EventBus.getInstance().unregister(this);
        super.onDestroy();
    }

    @Override
    public void onStart() {
        super.onStart();
        SearchView sw = (SearchView) this.getActivity().findViewById(R.id.toolbar).findViewById(R.id.searchView);
        sw.setQuery("", false);
        sw.setIconified(true);
        LayoutInflater inflater = this.getActivity().getLayoutInflater();
        View view = inflater.inflate(R.layout.loading, container, false);
        container.removeAllViews();
        container.addView(view);
        firstLoad = true;
        loadData();
    }

    public void onAsyncTaskResult(Event event) {
        searchRunning = false;
        LayoutInflater inflater = this.getActivity().getLayoutInflater();
        if (firstLoad) {
            View view = inflater.inflate(fragment_view, container, false);
            RecyclerView rv = (RecyclerView) view.findViewById(recycler_view);
            llm = new LinearLayoutManager(view.getContext());
            rv.setLayoutManager(llm);
            rv.setAdapter(adapter);
            if (endless) {
                rv.addOnScrollListener(new RecyclerView.OnScrollListener() {
                    @Override
                    public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                        visibleItemCount = llm.getChildCount();
                        totalItemCount = llm.getItemCount();
                        pastVisibleItems = llm.findFirstVisibleItemPosition();
                        if (!loading) {
                            if ((visibleItemCount + pastVisibleItems) >= totalItemCount - preloadCount) {
                                loading = true;
                                loadData();
                            }
                        }
                    }
                });
            }

            container.removeAllViews();
            container.addView(view);
            firstLoad = false;
        }
        result = event.getResult();
        if (!endless) {
            data.clear();
        }
        if (lastItem == null || !(lastItem.equals(result.get(0)))) {
            lastItem = result.get(0);
            data.addAll(event.getResult());
            adapter.notifyDataSetChanged();
            loading = false;
        }
    }
}

File: app/src/main/java/cz/kutner/comicsdb/fragment/AbstractFragment.kt
package cz.kutner.comicsdb.fragment

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import cz.kutner.comicsdb.R
import cz.kutner.comicsdb.Utils
import kotlinx.android.synthetic.fragment.*
import pl.aprilapps.switcher.Switcher
import rx.Subscription
import uk.co.ribot.easyadapter.EasyRecyclerAdapter
import java.util.*

public abstract class AbstractFragment<Item: Any> : Fragment() {
    var lastPage: Int = 0
    private var firstLoad: Boolean = false
    var searchRunning: Boolean = false
    private var loading: Boolean = false
    private var lastItem: Item? = null
    var data: MutableList<Item> = ArrayList()
    var result: List<Item> = ArrayList()
    private var pastVisibleItems: Int = 0
    private var visibleItemCount: Int = 0
    private var totalItemCount: Int = 0
    var adapter: EasyRecyclerAdapter<Any>? = null
    var preloadCount: Int = 0
    var endless: Boolean = false
    var spinnerEnabled: Boolean = false
    var spinnerValues: Array<String>? = null
    var filter: String
    private var spinnerPosition: Int? = null
    private var switcher: Switcher? = null

    var subscription: Subscription? = null

    init {
        lastPage = 1
        loading = false
        endless = true
        spinnerEnabled = false
        filter = ""
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater?.inflate(R.layout.fragment, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        switcher = Switcher.Builder().withContentView(content).withEmptyView(empty_view).withProgressView(progress_view).withErrorView(error_view).build()
        val llm = LinearLayoutManager(view?.context)
        recycler_view.layoutManager = llm
        recycler_view.adapter = adapter
        try_again.setOnClickListener {
            if (Utils.isConnected()) {
                checkConnectionAndLoadData();
            }
        }

        if (endless) {
            recycler_view.addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
                    visibleItemCount = llm.childCount
                    totalItemCount = llm.itemCount
                    pastVisibleItems = llm.findFirstVisibleItemPosition()
                    if (!loading) {
                        if ((visibleItemCount + pastVisibleItems) >= totalItemCount - preloadCount) {
                            loading = true
                            loadData()
                        }
                    }
                }
            })
        }
        switcher?.showProgressView()
    }

    abstract fun loadData()

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        checkConnectionAndLoadData()
    }

    private fun checkConnectionAndLoadData() {
        if (!Utils.isConnected()) {
            switcher?.showErrorView()
        } else {
            switcher?.showProgressView()
            firstLoad = true
            if (!result.isEmpty()) {
                showData()
            } else {
                loadData()
            }

        }
    }

    public fun showData() {
        searchRunning = false
        if (firstLoad) {
            if (!spinnerEnabled && spinner != null) {
                spinner?.visibility = View.GONE
                filter_text.visibility = View.GONE
            }
            if (spinnerEnabled && spinner != null) {
                val spinnerAdapter = ArrayAdapter(this.activity, android.R.layout.simple_spinner_item, spinnerValues)
                spinner?.adapter = spinnerAdapter
                spinnerAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
                if (spinnerPosition != null) {
                    spinner?.setSelection(spinnerPosition!!)
                } else {
                    spinnerPosition = 0
                }
                spinner?.onItemSelectedListener = itemSelectedListener()
            }
            switcher?.showContentView()
            firstLoad = false
        }
        if (result.size() > 0) {
            if (lastItem == null || lastItem != result.get(0)) {
                lastItem = result.get(0)
                if (!endless) {
                    data.clear()
                }
                data.addAll(result)
            }
        }
        adapter?.notifyDataSetChanged()
        loading = false
    }

    override fun onStop() {
        super.onStop()
        if (subscription != null) {
            subscription?.unsubscribe()
        }
    }

    private inner class itemSelectedListener : AdapterView.OnItemSelectedListener {

        override fun onItemSelected(parent: AdapterView<*>, view: View, pos: Int, id: Long) {
            if (spinnerPosition !== pos) {
                filter = spinner?.selectedItem.toString()
                data.clear()
                lastItem = null
                switcher?.showProgressView()
                firstLoad = true
                lastPage = 1
                spinnerPosition = pos
                loadData()
            }
        }

        override fun onNothingSelected(parent: AdapterView<*>) {

        }
    }
}